File: cpp/day1/day1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

int main() {
    std::ifstream ifs("input.txt");
    int left;
    int right;
    std::vector<int> left_side;
    std::vector<int> right_side;
    while (ifs >> left >> right) {
        left_side.push_back(left);
        right_side.push_back(right);
    }
    std::sort(left_side.begin(), left_side.end());
    std::sort(right_side.begin(), right_side.end());
    int sum = 0;
    /*
    Part 1:
    for (int i = 0; i < left_side.size(); i++) {
        sum += abs(right_side[i] - left_side[i]);
    }
    */
    for (int i = 0; i < left_side.size(); i++) {
        int partial = 0;
        for (int j = 0; j < right_side.size(); j++) {
            if (left_side[i] == right_side[j]) {
                partial++;
            }
        }
        sum += partial * left_side[i];
    }
    std::cout << sum << std::endl;
    return 0;
}

File: go/day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	fp, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer fp.Close()
	scanner := bufio.NewScanner(fp)
	scanner.Split(bufio.ScanLines)
	var leftSide []int
	var rightSide []int
	for scanner.Scan() {
		entry := strings.Split(scanner.Text(), "   ")
		left, _ := strconv.Atoi(entry[0])
		right, _ := strconv.Atoi(entry[1])
		leftSide = append(leftSide, left)
		rightSide = append(rightSide, right)
	}
	slices.Sort(leftSide)
	slices.Sort(rightSide)
	sum := 0
	/*
		Part 1:
		for i := 0; i < len(leftSide); i++ {
			dist := rightSide[i] - leftSide[i]
			if dist < 0 {
				dist = -dist
			}
			sum += dist
		}
	*/
	for i := 0; i < len(leftSide); i++ {
		partial := 0
		for j := 0; j < len(rightSide); j++ {
			if leftSide[i] == rightSide[j] {
				partial++
			}
		}
		sum += partial * leftSide[i]
	}
	fmt.Println(sum)
}
